File: OnePIC/app/src/main/java/com/goldenratio/onepic/LoadModule/LoadResolver.kt
package com.goldenratio.onepic.LoadModule

import android.util.Log
import com.goldenratio.onepic.JpegViewModel
import com.goldenratio.onepic.AllinJPEGModule.Contents.Audio
import com.goldenratio.onepic.AllinJPEGModule.Contents.ContentAttribute
import com.goldenratio.onepic.AllinJPEGModule.Contents.Picture
import com.goldenratio.onepic.AllinJPEGModule.Contents.Text
import com.goldenratio.onepic.AllinJPEGModule.AiContainer
import kotlinx.coroutines.*
import java.io.IOException


class LoadResolver() {
    companion object{
        const val MARKER_SIZE = 2
        const val APP3_FIELD_LENGTH_SIZE = 2
        const val FIELD_SIZE = 4
        const val BURST_MODE_SIZE = 1
    }

    fun isAllinJpegFormat(sourceByteArray : ByteArray, APP3_startOffset: Int) : Boolean{
        return (sourceByteArray[APP3_startOffset+4] == 0x4D.toByte() &&  sourceByteArray[APP3_startOffset+5] == 0x43.toByte()
            && sourceByteArray[APP3_startOffset+6] == 0x46.toByte() ||
            sourceByteArray[APP3_startOffset+4] == 0x41.toByte() &&  sourceByteArray[APP3_startOffset+5] == 0x69.toByte()
            && sourceByteArray[APP3_startOffset+6] == 0x46.toByte()  )
    }


    /**
     * TODO All-in JPEG 포맷인지 구별 후 APP3 pos 리턴
     *
     * @param sourceByteArray 사진 파일의 바이너리 데이터
     * @return All-in JPEG 포맷이면 APP3 시작 pos 리턴, 그렇지 않으면 -1 리턴
     */
    fun findAPP3StartPos(sourceByteArray: ByteArray) : Int {
        // APP3 - ALL in JPEG 세그먼트의 시작 위치를 찾음
        var APP3_startOffset = 2
        while (APP3_startOffset < sourceByteArray.size - 1) {
            if (sourceByteArray[APP3_startOffset] == 0xFF.toByte() && sourceByteArray[APP3_startOffset + 1] == 0xE3.toByte()) {
               val isAllinJpegFormat =  isAllinJpegFormat(sourceByteArray, APP3_startOffset)
                //All in Format인지 확인 TODO("All-in 과 MC과 합쳐져 있음. 후에 MC 지워주기")
                if(isAllinJpegFormat)
                    return APP3_startOffset
                else {
                    // APP3 마커가 있지만 All in Format이 아님
                    return -1
                }
                //break`
            }
            APP3_startOffset++
        }
        // APP3 마커가 없음
        return -1
    }

    /**
     * TODO JPEG과 All-in JPEG 포맷 구별 후, 포맷에 따라 AiContainer를 채우는 함수 호출
     *      JPEG -> setBasicJPEG 호출
     *      All-in JPEG -> parsingAllinJPEG 호출
     * @param AiContainer 갱신할 AiConatainer 객체
     * @param sourceByteArray 사진 파일의 바이너리 데이터
     */
    suspend fun createAiContainer(
        AiContainer: AiContainer,
        sourceByteArray: ByteArray
    ) {
        Log.d("AiContainer", "createMCContainer() sourceByreArray.Size : ${sourceByteArray.size}")
        // APP3 세그먼트의 시작 위치를 찾음
        var APP3_startOffset = 0
        APP3_startOffset = findAPP3StartPos(sourceByteArray)
        // APP3 세그먼트를 찾지 못함
        if (APP3_startOffset == - 1) {
            try{
                // 일반 JPEG
                Log.d("AiContainer", "createMCContainer() 일반 JPEG 생성")
                Log.d("version3", "createMCContainer() 일반 JPEG 생성")
                createAiConaterInJPEG(AiContainer, sourceByteArray)

            }catch (e : IOException){
                Log.e("Aicontainer", "createMCContainer() Basic JPEG Parsing 불가")
            }
        }
        else {
            Log.d("version3", "createMCContainer() All-in JPEG 생성")
            Log.d("AiConainer", "createMCContainer() All-in JPEG 생성")
            createAiConainerInAllinJPEG(AiContainer, sourceByteArray)
        }

    }

    fun createAiConaterInJPEG(AiContainer: AiContainer, sourceByteArray: ByteArray){
        JpegViewModel.AllInJPEG = false
        AiContainer.setBasicJepg(sourceByteArray)
    }

    suspend fun createAiConainerInAllinJPEG(AiContainer: AiContainer,
                                            sourceByteArray: ByteArray){
        CoroutineScope(Dispatchers.IO).launch {
            // APP3 세그먼트의 시작 위치를 찾음
            var APP3_startOffset = 2
            APP3_startOffset = findAPP3StartPos(sourceByteArray)

            try {
                JpegViewModel.AllInJPEG = true

                val isBurstMode = sourceByteArray.get( APP3_startOffset + MARKER_SIZE + APP3_FIELD_LENGTH_SIZE + FIELD_SIZE)
                var burst_mode_size = 1
                if(isBurstMode.toInt() == 1){
                    AiContainer.isBurst = true
                    Log.d("version3", "파싱된 burstMode : ${isBurstMode}")
                }
                else if (isBurstMode.toInt() == 0){
                    AiContainer.isBurst = false
                    Log.d("version3", "파싱된 burstMode : ${isBurstMode}")
                } else{
                    AiContainer.isBurst = true
                    burst_mode_size = 0
                }


                // 1. ImageContent Pasrsing
                var imageContentStartOffset =
                    APP3_startOffset + MARKER_SIZE + APP3_FIELD_LENGTH_SIZE + FIELD_SIZE + burst_mode_size
                var imageContentInfoSize = ByteArraytoInt(sourceByteArray, imageContentStartOffset)
                var pictureList = async {
                    imageContentParsing(
                        AiContainer,
                        sourceByteArray,
                        sourceByteArray.copyOfRange(
                            imageContentStartOffset,
                            imageContentStartOffset + imageContentInfoSize
                        ),
                        isBurstMode.toInt()
                    )
                }
                AiContainer.imageContent.setContent(pictureList.await())

                // 2. TextContent Pasrsing
                var textContentStartOffset =
                    APP3_startOffset + MARKER_SIZE + APP3_FIELD_LENGTH_SIZE + FIELD_SIZE + BURST_MODE_SIZE+ imageContentInfoSize
                var textContentInfoSize = ByteArraytoInt(sourceByteArray, textContentStartOffset)
                if (textContentInfoSize > 0) {
                    var textList = textContentParsing(
                        AiContainer, sourceByteArray,
//                            sourceByteArray.copyOfRange(
//                                textContentStartOffset + 4,
//                                textContentStartOffset + 8 + textContentInfoSize
//                            )
                        sourceByteArray.copyOfRange(
                            textContentStartOffset,
                            textContentStartOffset + 4 + textContentInfoSize
                        )
                    )
                    AiContainer.textContent.setContent(textList)
                }

                // 3. AudioContent Pasrsing
                var audioContentStartOffset = textContentStartOffset + textContentInfoSize
                var audioContentInfoSize = ByteArraytoInt(sourceByteArray, audioContentStartOffset)
                Log.d("AudioModule", "audioContentInfoSize : ${audioContentInfoSize}")
                if (audioContentInfoSize > 0) {
                    var audioDataStartOffset =
                        ByteArraytoInt(sourceByteArray, audioContentStartOffset + 4)
                    Log.d("AudioModule", "audioDataStartOffset : ${audioDataStartOffset}")

                    var audioAttribute =
                        ByteArraytoInt(sourceByteArray, audioContentStartOffset + 8)
                    Log.d("AudioModule", "audioAttribute : ${audioAttribute}")

                    var audioDataLength =
                        ByteArraytoInt(sourceByteArray, audioContentStartOffset + 12)
                    Log.d("AudioModule", "audioDataLength : ${audioDataLength}")
                    if (audioDataLength > 0) {
                        var audioBytes = sourceByteArray.copyOfRange(
                            audioDataStartOffset + MARKER_SIZE,
                            audioDataStartOffset + audioDataLength
                        )
                        Log.d("AudioModule", "audioBytes : ${audioBytes.size}")
                        var audio = Audio(audioBytes, ContentAttribute.fromCode(audioAttribute))
                        AiContainer.audioContent.setContent(audio)
                        if(AiContainer.audioResolver !=null ){
                            AiContainer.audioResolver!!.saveByteArrToAacFile(
                                audio._audioByteArray!!,
                                "viewer_record"
                            )

                        }


                    }
                    // MCContainer.audioResolver.saveByteArrToAacFile(audioBytes)
                }
            } catch (e: IOException) {
                Log.e("MCcontainer", "MC JPEG Parsing 불가")
            }
            return@launch
        }
    }

    fun ByteArraytoInt(byteArray: ByteArray, stratOffset : Int): Int {
        var intNum :Int = ((byteArray[stratOffset].toInt() and 0xFF) shl 24) or
                ((byteArray[stratOffset+1].toInt() and 0xFF) shl 16) or
                ((byteArray[stratOffset+2].toInt() and 0xFF) shl 8) or
                ((byteArray[stratOffset+3].toInt() and 0xFF))
        return intNum
    }

    suspend fun imageContentParsing(AiContainer: AiContainer, sourceByteArray: ByteArray, imageInfoByteArray: ByteArray, isBurstMode : Int): ArrayList<Picture> = withContext(Dispatchers.Default) {
        var picture : Picture
        var pictureList : ArrayList<Picture> = arrayListOf()

        var startIndex = 0
        var imageContentSize = ByteArraytoInt(imageInfoByteArray, startIndex)
        startIndex++
        var imageCount = ByteArraytoInt(imageInfoByteArray, startIndex*4)
        startIndex++
        for(i in 0..imageCount -1){
            var offset = ByteArraytoInt(imageInfoByteArray, (startIndex*4))
            startIndex++
            var app1DataSize = ByteArraytoInt(imageInfoByteArray, (startIndex*4))
            startIndex++
            var size = ByteArraytoInt(imageInfoByteArray, startIndex*4)
            startIndex++
            var attribute = ByteArraytoInt(imageInfoByteArray, startIndex*4)
            startIndex++
            var embeddedDataSize = ByteArraytoInt(imageInfoByteArray, startIndex*4)
            startIndex++
            var embeddedData : ArrayList<Int> = arrayListOf()
            if (embeddedDataSize > 0){
                var curInt : Int = 0
                for(j in 0..embeddedDataSize/4 -1){
                    // 4개씩 쪼개서 Int 생성
                    curInt = ByteArraytoInt(imageInfoByteArray, startIndex*4)
                    embeddedData.add(curInt)
                    startIndex++
                }
            }
            if(i==0){
                val jpegBytes = sourceByteArray.copyOfRange(offset,  offset + size - 1)
                // Jpeg Meta 데이터 떼기
                var jpegMetaData = AiContainer.imageContent.extractJpegMeta(sourceByteArray.copyOfRange(offset,
                    offset + size -1), ContentAttribute.fromCode(attribute))
                AiContainer.setJpegMetaBytes(jpegMetaData)

                var app1Segment : ByteArray?
                if(isBurstMode == 1){
                    app1Segment = null
                }else{
                     app1Segment = AiContainer.imageContent.extractAPP1(jpegBytes)
                }
              //  val app1Segment = AiContainer.imageContent.extractAPP1(jpegBytes)
                val frame =async {
                    AiContainer.imageContent.extractFrame(jpegBytes,ContentAttribute.fromCode(attribute))
                }
                picture = Picture(offset, app1Segment, frame.await(), ContentAttribute.fromCode(attribute), embeddedDataSize, embeddedData)
                picture.waitForByteArrayInitialized()
            }else{
                if(i == 5){
                    Log.d("ddd","sss")
                }
                val app1Segment = sourceByteArray.copyOfRange(offset + 2, offset + 2 + app1DataSize)
                val imageData = sourceByteArray.copyOfRange(offset + 2 + app1DataSize, offset + 2 + app1DataSize + size)
                // picture 생성
                picture = Picture(offset, app1Segment, imageData, ContentAttribute.fromCode(attribute), embeddedDataSize, embeddedData)
                picture.waitForByteArrayInitialized()
            }
            pictureList.add(picture)
            Log.d("version3", picture.toString())
            Log.d("Load_Module", "picutureList size : ${pictureList.size}")
        }
        return@withContext pictureList
    }

    fun textContentParsing(AiContainer: AiContainer, sourceByteArray: ByteArray, textInfoByteArray: ByteArray) : ArrayList<Text>{
        var textList : ArrayList<Text> = arrayListOf()
        var startIndex = 0
        var textContentInfoSize = ByteArraytoInt(textInfoByteArray, startIndex)
        startIndex++
        var textCount = ByteArraytoInt(textInfoByteArray, startIndex * FIELD_SIZE)
        startIndex++
        for(i in 0..textCount -1){
            var offset = ByteArraytoInt(textInfoByteArray, startIndex * FIELD_SIZE)
            startIndex++
            var attribute = ByteArraytoInt(textInfoByteArray, startIndex * FIELD_SIZE)
            startIndex++
            var size = ByteArraytoInt(textInfoByteArray,  startIndex * FIELD_SIZE)
            startIndex++

            val charArray = CharArray(size/2) // 변환된 char 값들을 담을 배열
            if (size > 0){
                if(sourceByteArray[offset] == 0xFF.toByte() && sourceByteArray[offset+1] == 0x20.toByte()){
                    for(j in 0..size -1 step 2){
                        // 4개씩 쪼개서 Int 생성
                        val charValue = ((sourceByteArray[offset + MARKER_SIZE + j].toInt() and 0xFF) shl 8) or
                                ((sourceByteArray[offset + MARKER_SIZE + j+1].toInt() and 0xFF) shl 0)
                        charArray[j / 2] = charValue.toChar() // char로 변환 후 배열에 저장
                    }
                }
            }
            var string : String = String(charArray)
            var text = Text(string, ContentAttribute.fromCode(attribute))
            Log.d("text_parsing", text.toString())
            textList.add(text)
        }
        return textList
    }


}

File: Chorme_extension/contentScript/All-in-JPEG/LoadResolver.js
import ContentAttribute from "./content/contentType.js";
import Picture from "./content/Picture.js";
import Aiaudio from "./content/Aiaudio.js";
import Text from "./content/Text.js";

export default class LoadResolver {
    static MARKER_SIZE = 2;
    static APP3_FIELD_LENGTH_SIZE = 2;
    static FIELD_SIZE = 4;
    static BURST_MODE_SIZE = 1;
    
    async isAllinJPEG(sourceByreArray){
        var APP3_startOffset = 2
        APP3_startOffset =  await this.findAPP3StartPos(sourceByreArray)
        if(APP3_startOffset == -1)

            return false;
        else 
            return true;
    }


    async findAPP3StartPos(sourceByteArray) {
        let APP3_startOffset = 2;
        while (APP3_startOffset < sourceByteArray.length - 1) {
            if (sourceByteArray[APP3_startOffset] === 0xFF && sourceByteArray[APP3_startOffset + 1] === 0xE3) {
                if (
                    (sourceByteArray[APP3_startOffset + 4] === 0x4D && sourceByteArray[APP3_startOffset + 5] === 0x43 &&
                        sourceByteArray[APP3_startOffset + 6] === 0x46) ||
                    (sourceByteArray[APP3_startOffset + 4] === 0x41 && sourceByteArray[APP3_startOffset + 5] === 0x69 &&
                        sourceByteArray[APP3_startOffset + 6] === 0x46)
                ) {
                    return APP3_startOffset;
                } else {
                    return -1;
                }
            }
            APP3_startOffset++;
        }
        return -1;
    }

    async createAiContainer(AiContainer, sourceByteArray) {

        console.log(`createAiContainer() sourceByreArray.Size : ${sourceByteArray.length}`);
        
        const APP3_startOffset = await this.findAPP3StartPos(sourceByteArray);
        
        if (APP3_startOffset === -1) {
            try {
                console.log("createAiContainer() 일반 JPEG 생성");
                AiContainer.setBasicJepg(sourceByteArray);
            } catch (e) {
                console.error("createAiContainer() Basic JPEG Parsing 불가", e);
            }
        } else {
            try {
                console.log("createAiContainer() MC JPEG 생성");
                console.log(`createAiContainer() App3 Start Offset : ${APP3_startOffset}`);
                //JpegViewModel.AllInJPEG = true;
                
                const isBurstMode = sourceByteArray[LoadResolver.APP3_startOffset + LoadResolver.MARKER_SIZE + LoadResolver.APP3_FIELD_LENGTH_SIZE + LoadResolver.FIELD_SIZE];
            

                // 1. Image Content Parsing
                const imageContentStartOffset = APP3_startOffset + LoadResolver.MARKER_SIZE + LoadResolver.APP3_FIELD_LENGTH_SIZE + LoadResolver.FIELD_SIZE + LoadResolver.BURST_MODE_SIZE;
                const imageContentInfoSize = this.ByteArraytoInt(sourceByteArray, imageContentStartOffset);
                const pictureList = await this.imageContentParsing(
                    AiContainer,
                    sourceByteArray,
                    sourceByteArray.slice(
                        imageContentStartOffset,
                        imageContentStartOffset + imageContentInfoSize
                    )
                );
                AiContainer.imageContent.setContent(pictureList);


                // 2. Text Conent Parsing
                const textContentStartOffset = APP3_startOffset + LoadResolver.MARKER_SIZE + LoadResolver.APP3_FIELD_LENGTH_SIZE + LoadResolver.FIELD_SIZE + LoadResolver.BURST_MODE_SIZE + imageContentInfoSize;
                const textContentInfoSize = this.ByteArraytoInt(sourceByteArray, textContentStartOffset);
                if (textContentInfoSize > 0) {
                    const textList = await this.textContentParsing(
                        AiContainer,
                        sourceByteArray,
                        sourceByteArray.slice(
                            textContentStartOffset,
                            textContentStartOffset + textContentInfoSize
                        )
                    );
                    AiContainer.textContent.setContentFromTextList(textList);
                }
                
                // 3. Audio Content Parsing
                const audioContentStartOffset = textContentStartOffset + textContentInfoSize;
                const audioContentInfoSize = this.ByteArraytoInt(sourceByteArray, audioContentStartOffset);
                console.log(`audioContentInfoSize : ${audioContentInfoSize}`);
                if (audioContentInfoSize > 0) {
                    const audioDataStartOffset = this.ByteArraytoInt(sourceByteArray, audioContentStartOffset + 4);
                    console.log(`audioDataStartOffset : ${audioDataStartOffset}`);

                    const audioAttribute = this.ByteArraytoInt(sourceByteArray, audioContentStartOffset + 8);
                    console.log(`audioAttribute : ${audioAttribute}`);

                    const audioDataLength = this.ByteArraytoInt(sourceByteArray, audioContentStartOffset + 12);
                    console.log(`audioDataLength : ${audioDataLength}`);
                    
                    if (audioDataLength > 0) {
                        const audioBytes = sourceByteArray.slice(
                            audioDataStartOffset + LoadResolver.MARKER_SIZE,
                            audioDataStartOffset + audioDataLength
                        );
                        console.log(`audioBytes : ${audioBytes.length}`);
                        
                        //let audio = new Aiaudio(audioBytes, ContentAttribute.fromCode(audioAttribute))
                        
                        AiContainer.audioContent.setContent(audioBytes, ContentAttribute.fromCode(audioAttribute));
                        // AiContainer.audioResolver.saveByteArrToAacFile(audioBytes, "viewer_record");
                    }
                }

            } catch (e) {
                console.error("MC JPEG Parsing 불가", e);
            }
        }
    }

    ByteArraytoInt(byteArray, stratOffset) {
        return (
            (byteArray[stratOffset] & 0xff) * 0x1000000 +
            (byteArray[stratOffset + 1] & 0xff) * 0x10000 +
            (byteArray[stratOffset + 2] & 0xff) * 0x100 +
            (byteArray[stratOffset + 3] & 0xff)
        );
    }

    async imageContentParsing(AiContainer, sourceByteArray, imageInfoByteArray) {
        const pictureList = [];
        let startIndex = 0;

        startIndex++;
        const imageCount = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4);
        startIndex++;
    
        for (let i = 0; i < imageCount; i++) {
            const offset = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4); startIndex++;
            const metaDataSize = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4); startIndex++;
            const size = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4); startIndex++;
            const attribute = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4); startIndex++;
            const embeddedDataSize = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4); startIndex++;
            
            const embeddedData = [];
            if (embeddedDataSize > 0) {
                for (let j = 0; j < embeddedDataSize / 4; j++) {
                    const curInt = this.ByteArraytoInt(imageInfoByteArray, startIndex * 4);
                    embeddedData.push(curInt);
                    startIndex++;
                }
            }
    
            let picture;
            if (i === 0) {
                const jpegBytes = sourceByteArray.slice(offset, offset + size - 1);
                // Jpeg Meta 데이터 떼기
                const jpegMetaData = AiContainer.imageContent.extractJpegMeta(
                    sourceByteArray.slice(offset, offset + size - 1),
                    ContentAttribute.fromCode(attribute)
                );
                AiContainer.setJpegMetaBytes(jpegMetaData);
            
                const frame = await AiContainer.imageContent.extractFrame(jpegBytes, ContentAttribute.fromCode(attribute));
                picture = new Picture(offset, jpegMetaData, frame, ContentAttribute.fromCode(attribute), embeddedDataSize, embeddedData);
                await picture.waitForByteArrayInitialized();
            } else {
                const metaData = sourceByteArray.slice(offset + 2, offset + 2 + metaDataSize);
                const imageData = sourceByteArray.slice(offset + 2 + metaDataSize, offset + 2 + metaDataSize + size);
                picture = new Picture(offset, metaData, imageData, ContentAttribute.fromCode(attribute), embeddedDataSize, embeddedData);
                await picture.waitForByteArrayInitialized();
            }
    
            pictureList.push(picture);
            console.log("Image_Parsing:", picture);
            console.log(`Load_Module: picutureList size: ${pictureList.length}`);
        }
    
        return pictureList;
    }
    
    async textContentParsing(AiContainer, sourceByteArray, textInfoByteArray) {
        const textList = [];
        let startIndex = 0;
        const textContentInfoSize = this.ByteArraytoInt(textInfoByteArray, startIndex);
        startIndex++;
        const textCount = this.ByteArraytoInt(textInfoByteArray, startIndex * LoadResolver.FIELD_SIZE);
        startIndex++;
    
        for (let i = 0; i < textCount; i++) {
            const offset = this.ByteArraytoInt(textInfoByteArray, startIndex * LoadResolver.FIELD_SIZE); startIndex++;
            const attribute = this.ByteArraytoInt(textInfoByteArray, startIndex * LoadResolver.FIELD_SIZE); startIndex++;
            const size = this.ByteArraytoInt(textInfoByteArray, startIndex * LoadResolver.FIELD_SIZE); startIndex++;
            
            const charArray = new Array(size / 2); // 변환된 char 값들을 담을 배열
            if (size > 0 && sourceByteArray[offset] === 0xFF && sourceByteArray[offset + 1] === 0x20) {
                for (let j = 0; j < size; j += 2) {
                    const charValue = ((sourceByteArray[offset + LoadResolver.MARKER_SIZE + j] & 0xFF) << 8) |
                        (sourceByteArray[offset + LoadResolver.MARKER_SIZE + j + 1] & 0xFF);
                    charArray[j / 2] = String.fromCharCode(charValue); // char로 변환 후 배열에 저장
                }
            }
            const string = charArray.join('');
            const text = new Text(string, ContentAttribute.fromCode(attribute));
            console.log(`text_parsing: ${text.toString()}`);
            textList.push(text);
        }
    
        return textList;
    }
    
}
